zookeeper-server/src/main/java/org/apache/zookeeper/server/TxnLogProposalIterator.java

```diff
@@ -87,23 +87,16 @@ public void remove() {
      * transaction records
      */
     public void close() {
-        if (itr != null) {
-            try {
-                itr.close();
-            } catch (IOException ioe) {
-                LOG.warn("Error closing file iterator", ioe);
-            }
+        try {
+            itr.close();
+        } catch (IOException ioe) {
+            LOG.warn("Error closing file iterator", ioe);
         }
     }
 
-    private TxnLogProposalIterator() {
-    }
-
-    public TxnLogProposalIterator(TxnIterator itr) {
-        if (itr != null) {
-            this.itr = itr;
-            hasNext = (itr.getHeader() != null);
-        }
+    public TxnLogProposalIterator(final TxnIterator itr) {
+        this.itr = Objects.requireNonNull(itr);
+        this.hasNext = (itr.getHeader() != null);
```

Comment:
- AFAIK, `Objects.requireNonNull` will throw an `NPE` when `itr` is null, is the new code changes the logic? Actually, `itr` can be null

Reply:
@maoling: @belugabehr explains this in the description section of [ZOOKEEPER-3676](https://issues.apache.org/jira/browse/ZOOKEEPER-3676):

> The the code manually returns an emptyIterator  when an error occurs, but it's also possible to return an emptyIterator by passing 'null' to the TxnLogProposalIterator constructor.  This is a bit ambiguous... why allow both?  Null values 'suck' so I think it's better to just make sure that emptyIterator is returned where needed an throw NPE if a 'null' value is passed in.

Reply:
Makes sense to me.
I think it would also worth to cover this behavior with unit tests.